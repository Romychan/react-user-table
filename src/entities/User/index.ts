export {
  useFetchUsers,
  useFetchUser,
  useCreateUser,
  useDeleteUser,
  useUpdateUser,
} from './api/queries';
export { userHandlers } from './api/msw/userHandlers';
export type { UsersResponse, UserDTO } from './api/types';
export { MOCK_USER } from './api/__mocks__/mocks';

export { UserContextProvider, UserContext } from './model/context';
export { useUserStore } from './model/hooks';
export {
  TABLE_USERS_COLUMNS,
  DEFAULT_USERS_PARAMS,
  FORM_USERS_FIELDS,
} from './model/constants';
export type { User, UserQueryParams, UserForm } from './model/types';
export { UserFormSchema } from './model/schemas';

export { UserAvatar } from './ui/UserAvatar/UserAvatar';
