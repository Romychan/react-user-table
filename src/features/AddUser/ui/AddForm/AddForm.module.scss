.container {
}

.form {
  display: grid;
  grid-template-columns: repeat(1, minmax(0, 1fr));
  gap: 1.5rem;
}

.actions {
  display: flex;
  justify-content: flex-end;
  gap: 0.5rem;
  margin-top: 0.5rem;

  & button {
    max-width: 90px;
    width: 100%;
  }
}
