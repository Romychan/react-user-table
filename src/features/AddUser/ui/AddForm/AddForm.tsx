import { Controller, useForm } from 'react-hook-form';
import { zodResolver } from '@hookform/resolvers/zod';

import {
  FORM_USERS_FIELDS,
  UserForm,
  UserFormSchema,
  useCreateUser,
} from '~/entities/User';

import { FormControl } from '~/shared/ui/FormControl';
import { Input } from '~/shared/ui/Input';
import { Button } from '~/shared/ui/Button';
import { InputPhone } from '~/shared/ui/InputPhone';

import styles from './AddForm.module.scss';

interface IAddFormProps {
  /** A callback to close the `<Drawer />`, which will be called after clicking Cancel or Update */
  onComplete: () => void;
}

/** This component of the form is for adding a new user */
export const AddForm = ({ onComplete }: IAddFormProps) => {
  const { control, handleSubmit } = useForm<UserForm>({
    resolver: zodResolver(UserFormSchema),
  });

  const { mutate: createUser } = useCreateUser();

  const onSubmit = (data: UserForm) => {
    onComplete();
    createUser(data);
  };

  return (
    <div className={styles.container}>
      <form className={styles.form} onSubmit={handleSubmit(onSubmit)}>
        {Object.entries(FORM_USERS_FIELDS).map(([field, label]) => {
          if (field === 'phone') {
            return (
              <Controller
                key={field}
                name={field}
                control={control}
                render={({
                  field: { name, value, onChange },
                  fieldState: { error },
                }) => (
                  <FormControl
                    label={label}
                    labelId={name}
                    error={error?.message}
                  >
                    <InputPhone
                      id={name}
                      name={name}
                      phoneCode="7"
                      value={value}
                      onChange={onChange}
                      error={!!error}
                      variant="filled"
                    />
                  </FormControl>
                )}
              />
            );
          }

          return (
            <Controller
              key={field}
              name={field as keyof UserForm}
              control={control}
              render={({
                field: { name, onChange },
                fieldState: { error },
              }) => (
                <FormControl
                  label={label}
                  labelId={name}
                  error={error?.message}
                >
                  <Input
                    id={name}
                    name={name}
                    onChange={onChange}
                    error={!!error}
                    variant="filled"
                  />
                </FormControl>
              )}
            />
          );
        })}

        <div className={styles.actions}>
          <Button
            type="button"
            size="sm"
            variant="secondary"
            data-testid="cancel-add-user"
            onClick={onComplete}
          >
            Cancel
          </Button>
          <Button type="submit" size="sm" data-testid="add-user">
            Add
          </Button>
        </div>
      </form>
    </div>
  );
};
