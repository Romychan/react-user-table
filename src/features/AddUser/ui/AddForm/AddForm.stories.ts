import type { Meta, StoryObj } from '@storybook/react';

import {
  withAppProvidersDecorator,
  withWidthDecorator,
} from '~/shared/lib/storybook';

import { AddForm } from './AddForm';

const meta = {
  title: 'Features/AddUser/AddForm',
  component: AddForm,
  parameters: {
    layout: 'centered',
    width: '420px',
  },
  decorators: [withAppProvidersDecorator, withWidthDecorator],
  tags: ['autodocs'],
} satisfies Meta<typeof AddForm>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {};
