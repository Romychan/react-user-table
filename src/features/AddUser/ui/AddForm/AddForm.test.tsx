import {
  fireEvent,
  screen,
  render,
  waitFor,
  RenderResult,
} from '@testing-library/react';
import { Mock, vi } from 'vitest';

import * as UserEntity from '~/entities/User';

import { AddForm } from './AddForm';

describe('AddForm', () => {
  let mockMutate: Mock<never, never>;
  let handleComplete: Mock<never, never>;
  let container: RenderResult;

  beforeEach(() => {
    mockMutate = vi.fn();
    handleComplete = vi.fn();

    vi.spyOn(UserEntity, 'useCreateUser').mockImplementation(
      vi.fn().mockReturnValue({
        mutate: mockMutate,
      }),
    );

    container = render(<AddForm onComplete={handleComplete} />);
  });

  it('should render correctly', () => {
    expect(screen.getAllByRole('textbox')).toHaveLength(6);
    expect(screen.queryAllByRole('alert')).toHaveLength(0);
    expect(screen.getByTestId('add-user')).toBeInTheDocument();
    expect(screen.getByTestId('cancel-add-user')).toBeInTheDocument();
  });

  it('should handle clicking on the cancel button correctly', () => {
    const cancelButton = screen.getByTestId('cancel-add-user');

    fireEvent.click(cancelButton);

    expect(handleComplete).toHaveBeenCalled();
  });

  it('should display an error if not all fields have been filled in', async () => {
    fillFormInputs();

    fireEvent.change(screen.getByRole('textbox', { name: 'Name' }), {
      target: { value: '' },
    });

    fireEvent.change(screen.getByRole('textbox', { name: 'Username' }), {
      target: { value: '' },
    });

    fireEvent.submit(screen.getByTestId('add-user'));

    expect(await screen.findAllByRole('alert')).toHaveLength(2);
    expect(handleComplete).not.toHaveBeenCalled();
    expect(mockMutate).not.toHaveBeenCalled();
  });

  it('should display matching error when email and phone is invalid', async () => {
    fillFormInputs();

    fireEvent.change(screen.getByRole('textbox', { name: 'Email' }), {
      target: { value: '123' },
    });

    fireEvent.change(screen.getByRole('textbox', { name: 'Phone' }), {
      target: { value: '321' },
    });

    fireEvent.submit(screen.getByTestId('add-user'));

    expect(await screen.findAllByRole('alert')).toHaveLength(2);
    expect(handleComplete).not.toHaveBeenCalled();
    expect(mockMutate).not.toHaveBeenCalled();
  });

  it('should not display error when value is valid', async () => {
    fillFormInputs();

    fireEvent.submit(screen.getByTestId('add-user'));

    await waitFor(() => expect(screen.queryAllByRole('alert')).toHaveLength(0));

    expect(handleComplete).toHaveBeenCalledTimes(1);
    expect(mockMutate).toHaveBeenCalledTimes(1);
  });

  it('should match snapshot', () => {
    expect(container).toMatchSnapshot();
  });
});

const fillFormInputs = () => {
  fireEvent.change(screen.getByRole('textbox', { name: 'Name' }), {
    target: { value: 'Name' },
  });
  fireEvent.change(screen.getByRole('textbox', { name: /username/i }), {
    target: { value: 'Username' },
  });
  fireEvent.change(screen.getByRole('textbox', { name: 'Email' }), {
    target: { value: 'test@test.test' },
  });
  fireEvent.change(screen.getByRole('textbox', { name: 'Phone' }), {
    target: { value: '+7 (123) 456-78-91' },
  });
  fireEvent.change(screen.getByRole('textbox', { name: 'City' }), {
    target: { value: 'City' },
  });
  fireEvent.change(screen.getByRole('textbox', { name: 'Address' }), {
    target: { value: 'Address' },
  });
};
