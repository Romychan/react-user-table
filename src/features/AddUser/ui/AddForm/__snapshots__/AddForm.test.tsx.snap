// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AddForm > should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="_container_990c92"
      >
        <form
          class="_form_990c92"
        >
          <div
            class="_container_77a9ed _primary_77a9ed _top_77a9ed"
          >
            <label
              class="_label_77a9ed"
              for="name"
            >
              Name
            </label>
            <div
              class="_content_77a9ed"
            >
              <input
                autocomplete="off"
                class="_input_95237c _filled_95237c"
                data-testid="input-field"
                id="name"
                name="name"
                value=""
              />
            </div>
          </div>
          <div
            class="_container_77a9ed _primary_77a9ed _top_77a9ed"
          >
            <label
              class="_label_77a9ed"
              for="username"
            >
              Username
            </label>
            <div
              class="_content_77a9ed"
            >
              <input
                autocomplete="off"
                class="_input_95237c _filled_95237c"
                data-testid="input-field"
                id="username"
                name="username"
                value=""
              />
            </div>
          </div>
          <div
            class="_container_77a9ed _primary_77a9ed _top_77a9ed"
          >
            <label
              class="_label_77a9ed"
              for="email"
            >
              Email
            </label>
            <div
              class="_content_77a9ed"
            >
              <input
                autocomplete="off"
                class="_input_95237c _filled_95237c"
                data-testid="input-field"
                id="email"
                name="email"
                value=""
              />
            </div>
          </div>
          <div
            class="_container_77a9ed _primary_77a9ed _top_77a9ed"
          >
            <label
              class="_label_77a9ed"
              for="phone"
            >
              Phone
            </label>
            <div
              class="_content_77a9ed"
            >
              <input
                autocomplete="off"
                class="_input_95237c _filled_95237c"
                data-testid="input-field"
                id="phone"
                name="phone"
                type="tel"
                value=""
              />
            </div>
          </div>
          <div
            class="_container_77a9ed _primary_77a9ed _top_77a9ed"
          >
            <label
              class="_label_77a9ed"
              for="city"
            >
              City
            </label>
            <div
              class="_content_77a9ed"
            >
              <input
                autocomplete="off"
                class="_input_95237c _filled_95237c"
                data-testid="input-field"
                id="city"
                name="city"
                value=""
              />
            </div>
          </div>
          <div
            class="_container_77a9ed _primary_77a9ed _top_77a9ed"
          >
            <label
              class="_label_77a9ed"
              for="address"
            >
              Address
            </label>
            <div
              class="_content_77a9ed"
            >
              <input
                autocomplete="off"
                class="_input_95237c _filled_95237c"
                data-testid="input-field"
                id="address"
                name="address"
                value=""
              />
            </div>
          </div>
          <div
            class="_actions_990c92"
          >
            <button
              class="_button_0302f3 _secondary_0302f3 _sm_0302f3"
              data-testid="cancel-add-user"
              type="button"
            >
              Cancel
            </button>
            <button
              class="_button_0302f3 _primary_0302f3 _sm_0302f3"
              data-testid="add-user"
              type="submit"
            >
              Add
            </button>
          </div>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="_container_990c92"
    >
      <form
        class="_form_990c92"
      >
        <div
          class="_container_77a9ed _primary_77a9ed _top_77a9ed"
        >
          <label
            class="_label_77a9ed"
            for="name"
          >
            Name
          </label>
          <div
            class="_content_77a9ed"
          >
            <input
              autocomplete="off"
              class="_input_95237c _filled_95237c"
              data-testid="input-field"
              id="name"
              name="name"
              value=""
            />
          </div>
        </div>
        <div
          class="_container_77a9ed _primary_77a9ed _top_77a9ed"
        >
          <label
            class="_label_77a9ed"
            for="username"
          >
            Username
          </label>
          <div
            class="_content_77a9ed"
          >
            <input
              autocomplete="off"
              class="_input_95237c _filled_95237c"
              data-testid="input-field"
              id="username"
              name="username"
              value=""
            />
          </div>
        </div>
        <div
          class="_container_77a9ed _primary_77a9ed _top_77a9ed"
        >
          <label
            class="_label_77a9ed"
            for="email"
          >
            Email
          </label>
          <div
            class="_content_77a9ed"
          >
            <input
              autocomplete="off"
              class="_input_95237c _filled_95237c"
              data-testid="input-field"
              id="email"
              name="email"
              value=""
            />
          </div>
        </div>
        <div
          class="_container_77a9ed _primary_77a9ed _top_77a9ed"
        >
          <label
            class="_label_77a9ed"
            for="phone"
          >
            Phone
          </label>
          <div
            class="_content_77a9ed"
          >
            <input
              autocomplete="off"
              class="_input_95237c _filled_95237c"
              data-testid="input-field"
              id="phone"
              name="phone"
              type="tel"
              value=""
            />
          </div>
        </div>
        <div
          class="_container_77a9ed _primary_77a9ed _top_77a9ed"
        >
          <label
            class="_label_77a9ed"
            for="city"
          >
            City
          </label>
          <div
            class="_content_77a9ed"
          >
            <input
              autocomplete="off"
              class="_input_95237c _filled_95237c"
              data-testid="input-field"
              id="city"
              name="city"
              value=""
            />
          </div>
        </div>
        <div
          class="_container_77a9ed _primary_77a9ed _top_77a9ed"
        >
          <label
            class="_label_77a9ed"
            for="address"
          >
            Address
          </label>
          <div
            class="_content_77a9ed"
          >
            <input
              autocomplete="off"
              class="_input_95237c _filled_95237c"
              data-testid="input-field"
              id="address"
              name="address"
              value=""
            />
          </div>
        </div>
        <div
          class="_actions_990c92"
        >
          <button
            class="_button_0302f3 _secondary_0302f3 _sm_0302f3"
            data-testid="cancel-add-user"
            type="button"
          >
            Cancel
          </button>
          <button
            class="_button_0302f3 _primary_0302f3 _sm_0302f3"
            data-testid="add-user"
            type="submit"
          >
            Add
          </button>
        </div>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
